File: nse_scanner_ui/src/api/today.ts
import { api } from "./client"

export interface StatusInfo {
  stale: boolean
  latest_bar: string | null
  latest_cal: string | null
  validation_fails_7d: number
  regime: string
  sessions_behind: number | null
}

export interface PositionRow {
  trade_id: number
  symbol: string
  entry_date: string
  entry_price: number
  qty: number
  stop_price: number
  days_held: number | null
  pnl_pct: number | null
  r_multiple: number | null
  status: "HOLD" | "WATCH" | "REVIEW"
  reasons: string[]
  lifecycle_status: "OPEN" | "PARTIAL_PROFIT" | "PARTIAL_LOSS" | "CLOSED"
  remaining_qty: number
  partial_pnl_rupees: number | null
}

export interface QualityChecklistItem {
  label: string
  passed: boolean
}

export interface OpportunitySignal {
  symbol: string
  trigger_price: number
  l1_price: number | null
  l2_price: number | null
  l1_l2_distance_pct: number | null
  neckline: number | null
  depth_pct: number | null
  stop_suggested: number | null
  target_suggested: number | null
  bottom_at_sma: string | null
  sma_stack: string | null
  rs_rank_pct: number | null
  // Descriptive-only forward outcome since the signal's scan_date -- null for
  // today's own signal (no bar after it yet); populated once you browse T-1+.
  outcome_status: "SL hit" | "Target hit" | "Towards target" | "Open for trade" | null
  outcome_date: string | null
  pct_since_signal: number | null
  // Evidence-calibrated quality score (scoring.py) -- 1D only, null for 1W/1M
  // (no comparable backtest history to calibrate against yet).
  quality_score: number | null
  quality_score_max: number | null
  quality_checklist: QualityChecklistItem[] | null
  historical_hit_rate_pct: number | null
  historical_sample_size: number | null
  historical_thin: boolean | null
}

export interface OpportunityBlock {
  timeframe: "1d" | "1w" | "1m"
  built: boolean
  total_signals: number
  new_signals: OpportunitySignal[]
  already_tracked_count: number
}

export interface WatchlistNearRow {
  symbol: string
  close: number | null
  target_price: number | null
  neckline: number | null
}

export interface PnlSummary {
  today: number
  this_week: number
  this_month: number
  all_time: number
  unrealised: number
}

export interface EquityPoint {
  exit_date: string
  cum_pnl: number
}

export interface TodayResponse {
  status: StatusInfo
  positions: PositionRow[]
  total_open_pnl: number
  at_risk_count: number
  opportunities: OpportunityBlock[]
  pnl: PnlSummary
  equity_curve: EquityPoint[]
  watchlist_near_trigger: WatchlistNearRow[]
}

export const todayApi = {
  get: () => api.get<TodayResponse>("/api/today"),
  opportunities: (asOfDate?: string) => {
    const params = asOfDate ? `?as_of_date=${asOfDate}` : ""
    return api.get<OpportunityBlock[]>(`/api/today/opportunities${params}`)
  },
  opportunityDates: (timeframe: string) =>
    api.get<{ dates: string[] }>(`/api/today/opportunity-dates?timeframe=${timeframe}`),
}

File: nse_scanner/src/api/schemas/today.py
"""api/schemas/today.py — response shape for GET /api/today (one aggregate call, see routers/today.py)."""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel


class StatusInfo(BaseModel):
    stale: bool
    latest_bar: date | None
    latest_cal: date | None
    validation_fails_7d: int
    regime: str
    sessions_behind: int | None = None


class PositionRow(BaseModel):
    trade_id: int
    symbol: str
    entry_date: date
    entry_price: float
    qty: int
    stop_price: float
    close: float | None
    days_held: int | None
    pnl_pct: float | None
    pnl_rupees: float | None
    r_multiple: float | None
    mae_pct: float | None
    mfe_pct: float | None
    status: str  # HOLD | WATCH | REVIEW -- advisor guidance, not position lifecycle
    reasons: list[str]
    lifecycle_status: str  # OPEN | PARTIAL_PROFIT | PARTIAL_LOSS | CLOSED
    remaining_qty: int
    partial_pnl_rupees: float | None


class OpportunitySignal(BaseModel):
    symbol: str
    trigger_price: float
    l1_price: float | None
    l2_price: float | None
    l1_l2_distance_pct: float | None
    neckline: float | None
    depth_pct: float | None
    stop_suggested: float | None
    target_suggested: float | None
    bottom_at_sma: str | None
    sma_stack: str | None
    rs_rank_pct: float | None


class OpportunityBlock(BaseModel):
    timeframe: str  # 1d | 1w | 1m
    built: bool  # False only for 1w/1m when that features table has zero rows
    total_signals: int
    new_signals: list[OpportunitySignal]  # full list, untracked (not already held/watched)
    already_tracked_count: int


class WatchlistNearRow(BaseModel):
    symbol: str
    close: float | None
    target_price: float | None
    neckline: float | None


class PnlSummary(BaseModel):
    today: float
    this_week: float
    this_month: float
    all_time: float
    unrealised: float


class EquityPoint(BaseModel):
    exit_date: str
    cum_pnl: float


class TodayResponse(BaseModel):
    status: StatusInfo
    positions: list[PositionRow]
    total_open_pnl: float
    at_risk_count: int
    opportunities: list[OpportunityBlock]
    pnl: PnlSummary
    equity_curve: list[EquityPoint]
    watchlist_near_trigger: list[WatchlistNearRow]
